source/adapters/opencl/device.hpp

```diff
@@ -10,11 +10,97 @@
 #pragma once
 
 #include "common.hpp"
+#include "platform.hpp"
 
-namespace cl_adapter {
-ur_result_t getDeviceVersion(cl_device_id Dev, oclv::OpenCLVersion &Version);
+struct ur_device_handle_t_ {
+  using native_type = cl_device_id;
+  native_type Device;
+  ur_platform_handle_t Platform;
+  cl_device_type Type = 0;
+  ur_device_handle_t ParentDevice = nullptr;
+  std::atomic<uint32_t> RefCount = 0;
+  bool IsNativeHandleOwned = true;
 
-ur_result_t checkDeviceExtensions(cl_device_id Dev,
-                                  const std::vector<std::string> &Exts,
-                                  bool &Supported);
-} // namespace cl_adapter
+  ur_device_handle_t_(native_type Dev, ur_platform_handle_t Plat,
+                      ur_device_handle_t Parent)
+      : Device(Dev), Platform(Plat), ParentDevice(Parent) {
+    RefCount = 1;
+    if (Parent) {
+      Type = Parent->Type;
+    } else {
+      clGetDeviceInfo(Device, CL_DEVICE_TYPE, sizeof(cl_device_type), &Type,
+                      nullptr);
+    }
+  }
+
+  ~ur_device_handle_t_() {
+    if (ParentDevice && IsNativeHandleOwned) {
+      clReleaseDevice(Device);
+    }
+  }
+
+  uint32_t incrementReferenceCount() noexcept { return ++RefCount; }
+
+  uint32_t decrementReferenceCount() noexcept { return --RefCount; }
```

Comment:
Reference count and associated member functions is something we should provide a shared abstraction for. That's out of scope for this PR, though. Seeing this just reminded me.

Reply:
Yeah, good point, that would eliminate a lot of repetition.

Reply:
Something like this already exists in at least the Native CPU adapter. I'm currently figuring out where this should live as the laoder also has reference counted objects.